### PR TITLE
Add access to WebXR depth texture + add example

### PR DESCRIPTION
**Pull request by cabanier**
Some authors want to use the depth texture to add an effect using a custom shader.
This diff enhances the WebXR code to allow access to the texture and add a sample.

Thanks to @hybridherbst !

*This contribution is funded by [Meta](https://meta.com)*

### CODE REVIEW

**Comment by cabanier** [comment]
@Mugen87 is there a way to change an existing shader? I'd like to reuse MeshLambertMaterial by adding my helper functions and inserting the calculation of the edge pixels.

**Comment by gkjohnson** [comment]
> @Mugen87 is there a way to change an existing shader? I'd like to reuse MeshLambertMaterial by adding my helper functions and inserting the calculation of the edge pixels.

You can use `onBeforeCompile` and modify the existing shader with find and replace:

```js
material.onBeforeCompile = shader => {
  
  shader.uniforms.depthTex = { value: /* texture here */ };
  shader.fragmentShader = shader.fragmentShader
    .replace( /#include<color_fragment>/g, match => /* glsl */`
      
      // your code here
      ${ match }

    ` );

};
```

This should also address #27664

**Comment by hybridherbst** [comment]
If you want it globally included you can also replace things in `ShaderChunks.someSpecificChunk`, then it doesn't need to be modified in every single material.
What I'm not sure in that second approach is what the right way is to set some uniforms either for all objects or per object; @gkjohnson do you have a recommendation there?

**Comment by gkjohnson** [comment]
> If you want it globally included you can also replace things in ShaderChunks.someSpecificChunk, then it doesn't need to be modified in every single material.

Generally I'd discourage modifying the built in shaders in the examples. It works but it's brittle. And of course having multiple libraries or components modifying built in materials can easily lead to overwritten changes and breakage.

> What I'm not sure in that second approach is what the right way is to set some uniforms either for all objects or per object;

Global uniforms aren't supported so you'll have to set it per material. UBOs are now in core but the current implementation doesn't make this case much more ergonomic. See #16922.

Encapsulating the above in a class can make this easier to manage at an application level but I don't think it's necessary for the example:

```js
class CustomMaterial extends MeshStandardMaterial {
  set depthTex( v ) { 
    this._uniforms.depthTex.value = v;
  }
  constructor( ...args ) {
    super( ...args );
    this.onBeforeCompile = shader => {
      // ...
      this._uniforms = shader.uniforms;
    };
  }

}
```

**Comment by hybridherbst** [comment]
> having multiple libraries or components modifying built in materials 

The same is true for `Material.onBeforeCompile` which isn't an event listener but a single assigned method, or am I understanding that wrong?

I'm just not sure how someone could take that sample and extend it; a common flow would be "this is a cool sample, now I want to load a glTF and keep the materials in that and apply that fancy occlusion." – how would someone do that? Need to be super careful then to not "miss" any material to modify it, can't use other onBeforeCompile callbacks, etc.

**Comment by gkjohnson** [comment]
> The same is true for Material.onBeforeCompile which isn't an event listener but a single assigned method, or am I understanding that wrong?

You can design for this case into your app or design your wrapper class to take an onBeforeCompile function and call it if it's set (ie adding a getter / setter for `onBeforeCompile` to your class)

> I'm just not sure how someone could take that sample and extend it; a common flow would be "this is a cool sample, now I want to load a glTF and keep the materials in that and apply that fancy occlusion."

This is always the case if you want to use a different material on a loaded model. My understanding is that `onBeforeCompile` is the way to modify built in materials but don't get me wrong - I think it's more difficult than it should be right now (see my issues on override materials, global uniforms, etc). I'd like to see improvements but I think discussions around how to adjust the ergonomics should happen in a separate thread and not block this PR.

**Comment by cabanier** [comment]
@Mugen87 it doesn't look like I have access to the internal shaders to extend them. What do you think of this change?
Alternatively, I can just make the changes to the webxr manager class and leave it up to authors to use the depth sensing texture.

**Comment by Mugen87** [comment]
I would leave the PR in its current state then 👍 .

**Comment by odysseyjason** [comment]
Is it possible to query the depth texture per frame from the render-loop, to for instance bounce objects off of walls sensed through the depth.api?  (the [webxr cubes example]( https://threejs.org/examples/?q=webxr#webxr_xr_cubes)  currently has the cubes getting occluded behind walls, would it be possible to query the depth data to see if they should bounce off of them instead?)  

I have looked at the  code from the [webxr cubes example]( https://threejs.org/examples/?q=webxr#webxr_xr_cubes) and [WebXRDepthSensing.js](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webxr/WebXRDepthSensing.js) and[ WebXRManager.js](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webxr/WebXRManager.js) but could not find where it was being defined if this depth data was GPU optimized texture or CPU optimized or understand where one could interject real time interaction possibilities with the depth data.

Any example that could show interaction with the Depth api texture data would be most helpful!  Also, a CPU optimized depth texture example would be great!

**Comment by cabanier** [comment]
> Any example that could show interaction with the Depth api texture data would be most helpful! Also, a CPU optimized depth texture example would be great!

If you look at this PR, the `getDepthTexture` call returns you the depth texture. You can use it in your shader just like a regular WebGL texture.

**Comment by Mugen87** [comment]
@cabanier Do you mind rebasing this PR? Sorry for overlooking it. But now that depth sensing is fixed via #28530 the PR can be eventually merged.

**Comment by cabanier** [comment]
> @cabanier Do you mind rebasing this PR? Sorry for overlooking it. But now that depth sensing is fixed via #28530 the PR can be eventually merged.

I can do that!

**Comment by Mugen87** [comment]
It seems the E2E screenshot of the new example requires an update. Can you quickly regenerate the screenshot and see if it helps?

If regenerating on your side does not fix the failure, I'll try it myself after the merge.

**Comment by cabanier** [comment]
> It seems the E2E screenshot of the new example requires an update. Can you quickly regenerate the screenshot and see if it helps?

Would you mind regenerating it? I developer on a headless server and it doesn't have chrome.

**Comment by Mugen87** [comment]
> Would you mind regenerating it?

Np, see #28535.

File: examples/webxr_xr_dragging_custom_depth.html [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+			function render() {
+
+				if (renderer.xr.hasDepthSensing() && ! isDepthSupplied) {
+
+					group.children.forEach(child => {
+
+						child.material.uniforms.depthColor.value = renderer.xr.getDepthTexture();
+						child.material.uniforms.depthWidth.value = 1680;
+						child.material.uniforms.depthHeight.value = 1760;
+						child.material.uniforms.NeedUpdate = true;
```
Comment (linbingquan) [inline examples/webxr_xr_dragging_custom_depth.html:428]: I was search `NeedUpdate` keyword in `src`, the result show only show something about `material.uniformsNeedUpdate = false;`

`uniforms.NeedUpdate` should be `uniformsNeedUpdate`?

or maybe I miss something.

Comment (Mugen87) [reply examples/webxr_xr_dragging_custom_depth.html:428]: Good catch! `child.material.uniforms.NeedUpdate` can just be removed. Would you like to do that with a PR?

**Pull request merged**